### PR TITLE
PE-3581: Write custom metadata while syncing

### PR DESCRIPTION
**Pull request by matibat**
Changes
- Adds to drive, folder, and file entities the customJsonMetaData field
- Sets the value ONLY at sync time - For creation/rename/move and any other operation, it's not being set

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/5v4k4e780i600
iOS release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:ios:06ea33a95a2e0d42ffce07/releases/5osi9ig195g98

### CODE REVIEW

File: lib/services/arweave/arweave_service.dart [hunk 1]
```diff
@@ -221,40 +221,8 @@ class ArweaveService {
     required String ownerAddress,
     required SnapshotDriveHistory snapshotDriveHistory,
     required DriveID driveId,
+    required List<Uint8List> entitiesData,
   }) async {
-    // FIXME - PE-3440
```
Comment (matibat) [inline lib/services/arweave/arweave_service.dart:225]: Revert this unnecessary change

**Review by matibat — COMMENTED** [review]
Self-review checkpoint

File: lib/services/arweave/arweave_service.dart (hunk 1 above)
Comment (matibat) [reply lib/services/arweave/arweave_service.dart:225]: fa3d334337137bf664f9a6656e8dcb2afa295870

**Comment by matibat** [comment]
TODO: use logger instead of print

File: lib/entities/drive_entity.dart [hunk 2]
```diff
@@ -77,6 +82,14 @@ class DriveEntity extends Entity {
     }
   }
 
+  static String customMetaDataFromData(Map<String, dynamic> metadata) {
```
Comment (karlprieb) [inline lib/entities/drive_entity.dart:85]: how is the structure of `entityJson`? Just curious to understand how it works :)
To get the custom metadata we only need to discard `name` and `rootFolderId`? Is there any other property besides these that are not custom metadata?

File: lib/blocs/upload/upload_handles/file_v2_upload_handle.dart [hunk 3]
```diff
@@ -52,7 +52,10 @@ class FileV2UploadHandle implements UploadHandle {
     await driveDao.transaction(() async {
       await driveDao.writeFileEntity(entity, path);
       await driveDao.insertFileRevision(
-        entity.toRevisionCompanion(performedAction: revisionAction),
+        entity.toRevisionCompanion(
+          performedAction: revisionAction,
+          customMetaData: null,
```
Comment (karlprieb) [inline lib/blocs/upload/upload_handles/file_v2_upload_handle.dart:57]: Is it `null` because we don't handle customMetaData on the file creation yet, right?

File: lib/entities/drive_entity.dart [hunk 4]
```diff
@@ -77,6 +82,14 @@ class DriveEntity extends Entity {
...
+    metadata.remove('name');
+    metadata.remove('rootFolderId');
+    final customMetadataAsString = json.encode(metadata);
+    print('Custom metadata for drive: $customMetadataAsString');
```
Comment (karlprieb) [inline lib/entities/drive_entity.dart:89]: as you already pointed in a previous comment let's use logger

File: lib/entities/file_entity.dart [hunk 5]
```diff
@@ -81,12 +85,24 @@ class FileEntity extends Entity {
...
   }
 
+  static String customMetaDataFromData(Map<String, dynamic> metadata) {
+    metadata.remove('name');
+    metadata.remove('size');
+    metadata.remove('lastModifiedDate');
+    metadata.remove('dataTxId');
+    metadata.remove('dataContentType');
+    final customMetadataAsString = json.encode(metadata);
+    print('Custom metadata for file: $customMetadataAsString');
```
Comment (karlprieb) [inline lib/entities/file_entity.dart:102]: as you already pointed in a previous comment let's use logger

File: lib/services/arweave/arweave_service.dart [hunk 6]
```diff
@@ -248,15 +250,13 @@ class ArweaveService {
 
           return _getEntityData(
             entityId: entity.id,
-            driveId: driveId,
+            driveId: ownerAddress,
```
Comment (karlprieb) [inline lib/services/arweave/arweave_service.dart:253]: `ownerAddress`?

Comment (matibat) [reply lib/services/arweave/arweave_service.dart]: Nice catch - 45c3f4f7d231fd1c1d6728568bb275f7dfaea622

File: lib/entities/drive_entity.dart (hunk 4 above)
Comment (matibat) [reply lib/entities/drive_entity.dart]: 56f5ee232bccb0813e47abbf32645fd6eadb16e5

File: lib/entities/file_entity.dart [hunk 7]
```diff
@@ -81,12 +85,24 @@ class FileEntity extends Entity {
...
         ..ownerAddress = transaction.owner.address
         ..bundledIn = transaction.bundledIn?.id
+        ..customJsonMetaData = customMetaDataFromData(entityJson)
         ..createdAt = commitTime;
     } catch (_) {
       throw EntityTransactionParseException(transactionId: transaction.id);
     }
   }
 
+  static String customMetaDataFromData(Map<String, dynamic> metadata) {
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/entities/file_entity.dart]: Declares a factory function - b4b522d0a8cf6ca0d658101ee7953355cd557709
Unit tests the factory and the implementations - 169d1a35b3d150fd32b214d93448a015c281d9a7

File: lib/entities/drive_entity.dart (hunk 2 above)
Comment (karlprieb) [reply lib/entities/drive_entity.dart]: Thanks :D

File: lib/entities/custom_metadata.dart [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import 'dart:convert';
+
+String Function() customMetadataFactory(
+  Map<String, dynamic> metadata,
+  List<String> reservedFields,
+) {
+  return () {
+    metadata.removeWhere((key, value) => reservedFields.contains(key));
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/entities/custom_metadata.dart:11]: There: c1405d204753989f62f2e4d8108690b8c5d73066

File: lib/entities/custom_metadata.dart [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import 'package:ardrive/utils/logger/logger.dart';
+
+const Map<String, List<String>> reservedFieldsPerEntityType = {
+  'drive': ['name', 'rootFolderId'],
+  'folder': ['name'],
+  'file': ['name', 'size', 'lastModifiedDate', 'dataTxId', 'dataContentType'],
+  'test': ['reserved'],
+};
+
+String Function() customMetadataFactory(
```
Comment (karlprieb) [inline lib/entities/custom_metadata.dart:12]: call it `extractCustomMetadataFromEntity` and call it directly on the places you need it.

Comment (matibat) [reply lib/entities/custom_metadata.dart]: Makes the method not to be a factory anymore; removes the static method customMetaDataFromData from entities: 55092b469badd4ad2c2cd26a8f5c3be48558d66b
Updates unit tests: 5db32fa4c3e272ed27dd9c275595a0719b7b8482

**Review by karlprieb — APPROVED** [review]
